02_C# Fundamentals/FileSystemApp/FileSystemApp/FileSystemVisitor.cs

```diff
@@ -0,0 +1,128 @@
+    {
+        public int NumberOfDirectories { get; set; }
+        public int NumberOfFilteredDirectories { get; set; }
+        public int NumberOfFiles { get; set; }
+        public int NumberOfFilteredFiles { get; set; }
+
+        bool includeDirectory;
+        bool includeFile;
+
+        public event EventHandler<OutputMessageEventArgs> StartMessage;
+        public void StartMessageFire()
+        {
+            StartMessage?.Invoke(this, new OutputMessageEventArgs
+            {
+                Message = "-------------Searching was started-------------"
+            });
+        }
+
+        public event EventHandler<OutputMessageEventArgs> EndMessage;
+        public void EndMessageFire()
+        {
+            EndMessage?.Invoke(this, new OutputMessageEventArgs
+            {
+                Message = "-------------Searching was finished-------------"
+            });
+        }
+
+        public event EventHandler<OutputMessageEventArgs> FileFound;
+        public event EventHandler<OutputMessageEventArgs> FilteredFileFound;
+        public event EventHandler<OutputMessageEventArgs> DirectoryFound;
+        public event EventHandler<OutputMessageEventArgs> FilteredDirectoryFound;
+
+        public Predicate<string> Filter { get; set; }
+
+        public FileSystemVisitor()
+        {
+
+        }
+
+        public FileSystemVisitor(Predicate<string> filter)
+        {
+            Filter = filter;
+        }
+
+        public IEnumerable<string> GetAllFoldersAndFiles(string startAddress)
+        {
+            if (!string.IsNullOrEmpty(startAddress) && Directory.Exists(startAddress))
+            {
+                var entities = FindEntities(startAddress);
+
+                foreach (var item in entities)
+                {
+                    if (Directory.Exists(item))
+                    {
+                        ++NumberOfDirectories;
+
+                        DirectoryFound?.Invoke(this, new OutputMessageEventArgs
+                        {
+                            Message = "Directory was found: " + Path.GetFileName(item),
+                            NumberOfDirectories = NumberOfDirectories
+                        });
+
+                        if (Filter != null)
+                        {
+                            if (!Filter(item))
+                            {
+                                continue;
+                            }
+
+                            ++NumberOfFilteredDirectories;
+
+                            FilteredDirectoryFound?.Invoke(this, new OutputMessageEventArgs
+                            {
+                                Message = "Filtered directory was found: " + Path.GetFileName(item),
+                                NumberOfDirectories = NumberOfFilteredDirectories
+                            });
+                        }
+                        yield return item;
+                    }
+
+                    if (File.Exists(item))
```

Comment:
Try to refactor this method, so that it doesn't look like an obvious code duplication at least. For example, you can at least move handling of File and Directory into a separate methods with comprehensive names, and call them from this method.

Reply:
done